File: app/src/main/java/com/hoymm/root/tictactoe2/GameEngine/GameBoardFragment.java
package com.hoymm.root.tictactoe2.GameEngine;

import android.content.res.Resources;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.airbnb.lottie.LottieAnimationView;
import com.hoymm.root.tictactoe2.R;

/**
 * Created by hoymm on 03.11.17.
 */

public class GameBoardFragment extends Fragment {
    private static final int fieldMargin = dpToPx(10);
    private static final int fieldPadding = dpToPx(10);
    private BoardSize boardSize;


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.game_board_fragment, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);


        boardSize = (BoardSize) getArguments().get(GameEngine.Companion.getGAME_BOARD_SIZE_KEY());
        view.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            public void onGlobalLayout() {
                getView().getViewTreeObserver().removeGlobalOnLayoutListener(this);
                generateGameBoard();
            }
        });


    }

    public static int dpToPx(int dp)
    {
        return (int) (dp * Resources.getSystem().getDisplayMetrics().density);
    }

    private void generateGameBoard() {
        LinearLayout generalLinearLayout = getActivity().findViewById(R.id.game_board_fragment_id);

        int howManyFieldsInRow = getHowManyFieldsInRow();
        int boardLength = Math.min(generalLinearLayout.getWidth(), generalLinearLayout.getHeight());
        boardLength -= generalLinearLayout.getPaddingLeft()*2;

        FieldParameters fieldParameters = new FieldParameters();
        fieldParameters.fieldLength = boardLength/howManyFieldsInRow - fieldMargin;

        for (int row = 0; row < howManyFieldsInRow; ++row) {
            boolean isItBottomRow = row == howManyFieldsInRow-1;
            fieldParameters.setMarginsToZero();
            setVerticalMargins(isItBottomRow, fieldParameters);
            insertBoardRow(generalLinearLayout, howManyFieldsInRow, fieldParameters);
        }
    }

    private void setVerticalMargins(boolean isItBottomRow, FieldParameters fieldParameters) {
        fieldParameters.marginBottom = fieldMargin;
        if (isItBottomRow)
            fieldParameters.marginBottom = 0;
    }

    private class FieldParameters {
        int fieldLength;
        int marginLeft, marginTop, marginRight, marginBottom;

        void setMarginsToZero(){
            marginLeft = marginTop = marginRight = marginBottom = 0;
        }
    }

    private int getHowManyFieldsInRow() {
        switch (boardSize){
            default:
            case board3x3:
                return 3;
            case board5x5:
                return 5;
            case board7x7:
                return 7;
        }
    }

    private void insertBoardRow(LinearLayout generalLinearLayout, int howManyFieldsInRow, FieldParameters fieldParameters) {
        LinearLayout rowLinearLayout = createRowLinearLayout(fieldParameters);
        generalLinearLayout.addView(rowLinearLayout);
        addGameFieldsToFillRow(rowLinearLayout, howManyFieldsInRow, fieldParameters);
    }

    private LinearLayout createRowLinearLayout(FieldParameters fieldParameters) {
        LinearLayout rowLinearLayout = new LinearLayout(getContext());
        LinearLayout.LayoutParams rowParams = new LinearLayout.LayoutParams
                (fieldParameters.fieldLength*getHowManyFieldsInRow() + (getHowManyFieldsInRow()-1)*fieldMargin
                        , fieldParameters.fieldLength);
        rowLinearLayout.setLayoutParams(rowParams);
        rowLinearLayout.setGravity(Gravity.CENTER);
        rowLinearLayout.setOrientation(LinearLayout.HORIZONTAL);
        rowLinearLayout.setBackgroundColor(Color.RED);
        return rowLinearLayout;
    }

    private void addGameFieldsToFillRow(LinearLayout rowLinearLayout, int howManyFieldsInRow, FieldParameters fieldParameters) {
        for (int fieldIndex = 0; fieldIndex < howManyFieldsInRow; ++fieldIndex) {
            boolean isItLastRightField = fieldIndex == howManyFieldsInRow-1;
            setHorizontalMargins(isItLastRightField, fieldParameters);
            rowLinearLayout.addView(createNewBoardField(fieldParameters));
        }
    }

    private void setHorizontalMargins(boolean isItLastField, FieldParameters fieldParameters) {
        fieldParameters.marginRight = fieldMargin;
        if (isItLastField)
            fieldParameters.marginRight = 0;
    }

    private LottieAnimationView createNewBoardField(FieldParameters fieldParameters) {
        LottieAnimationView myLottieAnimationView = new LottieAnimationView(getContext());

        LinearLayout.LayoutParams fieldParams = new LinearLayout.LayoutParams(
                fieldParameters.fieldLength, fieldParameters.fieldLength);
        fieldParams.leftMargin = fieldParameters.marginLeft;
        fieldParams.topMargin = fieldParameters.marginTop;
        fieldParams.rightMargin = fieldParameters.marginRight;
        fieldParams.bottomMargin = fieldParameters.marginBottom;

        myLottieAnimationView.setLayoutParams(fieldParams);
        myLottieAnimationView.setPadding(fieldPadding, fieldPadding, fieldPadding, fieldPadding);
        //myLottieAnimationView.setAdjustViewBounds(true);
        myLottieAnimationView.setScaleType(ImageView.ScaleType.CENTER_CROP);
        myLottieAnimationView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.appBackground));


        myLottieAnimationView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ((LottieAnimationView)v).setAnimation("cross.json");
                ((LottieAnimationView)v).playAnimation();
            }
        });

        return myLottieAnimationView;
    }

    public void setCircleTurnNow() {

    }
}

File: app/src/main/java/com/hoymm/root/tictactoe2/GameEngine/GameBoardFragment.kt
package com.hoymm.root.tictactoe2.GameEngine

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.*
import android.widget.GridView
import android.widget.LinearLayout
import com.hoymm.root.tictactoe2.R
import java.lang.ClassCastException

class GameBoardFragment : Fragment(), CheckIsGameFinished {
    private lateinit var gameBoard: GridView
    private var boardSize: BoardSize? = null
    private var fieldLength : Int = 0
    private var fieldsSeparatorLength: Int = 0
    private lateinit var currentAppDataInfo : CurrentAppDataInfo

    private val howManyFieldsInRow: Int get() =
       when (boardSize) {
            BoardSize.board3x3 -> 3
            BoardSize.board5x5 -> 5
            BoardSize.board7x7 -> 7
            else -> 3
        }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        try{
            currentAppDataInfo = context as CurrentAppDataInfo
        }
        catch (exception : ClassCastException){
            throw ClassCastException(activity.toString() + " must implement CurrentAppDataInfo");
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View?
            = inflater!!.inflate(R.layout.game_board_fragment, container, false)

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        boardSize = arguments.get(GameEngine.GAME_BOARD_SIZE_KEY) as BoardSize

        view!!.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                getView()!!.viewTreeObserver.removeOnGlobalLayoutListener(this)
                generateGameBoard()
            }
        })
    }

    private fun generateGameBoard() {
        gameBoard = activity.findViewById<GridView>(R.id.game_board_fragment_id)
        calculateFieldAndSeparatorLength()
        setNewWidthAndHeightForGameBoardView(gameBoard)
        setAdapter(gameBoard)
        setNumOfColumnsAndSpacingGridViewShouldHave(gameBoard)
    }

    private fun calculateFieldAndSeparatorLength() {
        val parentLayoutShorterSideLength =  Math.min(getParentLinearLayout().height, getParentLinearLayout().width)
        fieldsSeparatorLength = parentLayoutShorterSideLength/((Math.sqrt(howManyFieldsInRow.toDouble())*50).toInt())
        fieldLength = getTotalLengthOfFieldsInRow(parentLayoutShorterSideLength)/howManyFieldsInRow
    }

    private fun setNewWidthAndHeightForGameBoardView(gameBoardGridView: GridView) {
        val gameBoardSideLength = getTotalLengthOfSeparatorsInRow() + fieldLength * howManyFieldsInRow
        gameBoardGridView.layoutParams.width = gameBoardSideLength
        gameBoardGridView.layoutParams.height = gameBoardSideLength
    }

    private fun setAdapter(gameBoardLayout: GridView) {
        gameBoardLayout.adapter = GameBoardAdapter(activity, howManyFieldsInRow, fieldLength)
    }

    private fun setNumOfColumnsAndSpacingGridViewShouldHave(gameBoardLayout: GridView) {
        gameBoardLayout.numColumns = howManyFieldsInRow
        gameBoardLayout.horizontalSpacing = fieldsSeparatorLength
        gameBoardLayout.verticalSpacing = fieldsSeparatorLength
    }

    private fun getParentLinearLayout() =
            activity.findViewById<LinearLayout>(R.id.game_board_background_layout)

    private fun getTotalLengthOfFieldsInRow(shorterParentSideLength: Int) =
            shorterParentSideLength - getTotalLengthOfSeparatorsInRow()

    private fun getTotalLengthOfSeparatorsInRow() =
            howManyFieldSeparatorsInRow() * fieldsSeparatorLength

    private fun howManyFieldSeparatorsInRow() = (howManyFieldsInRow - 1)

    override fun checkIfSomeoneWon(): Shape? {
        var whoWon: Shape?
        val winCheck = CheckWin(gameBoard)

        whoWon = winCheck.checkIfWinInRows(getHowManyPointsWins())
        if (whoWon != null) return whoWon

        whoWon = winCheck.checkIfWinInColumns(getHowManyPointsWins())
        if (whoWon != null) return whoWon

        whoWon = winCheck.checkIfWinDiagonalFromTopRow(getHowManyPointsWins())
        if (whoWon != null) return whoWon

        whoWon = winCheck.checkIfWinDiagonalLeftColumnToBottom(getHowManyPointsWins())
        if (whoWon != null) return whoWon

        return null
    }

    override fun checkIfItIsADraw(): Boolean {
        var isItADraw: Boolean?
        val drawCheck = CheckDraw(gameBoard)

        isItADraw = drawCheck.checkIfDrawInRows(getHowManyPointsWins())
        if (isItADraw == false) return false

        isItADraw = drawCheck.checkIfDrawInColumns(getHowManyPointsWins())
        if (isItADraw == false) return false

        isItADraw = drawCheck.checkIfDrawDiagonalFromTopRow(getHowManyPointsWins())
        if (isItADraw == false) return false

        isItADraw = drawCheck.checkIfDrawDiagonalLeftColumnToBottom(getHowManyPointsWins())
        if (isItADraw == false) return false

        return true
    }

    private fun getHowManyPointsWins() = if (boardSize == BoardSize.board3x3) 3 else 4
}
